src/daemon/rpmostreed-os.cxx

```diff
@@ -1138,6 +1140,92 @@ os_handle_get_packages (RPMOSTreeOS *interface, GDBusMethodInvocation *invocatio
   return TRUE;
 }
 
+static void
+query_results_to_builder (HyQuery query, GVariantBuilder *builder, const gchar *id)
+{
+  g_autoptr (GPtrArray) pkglist = NULL;
+  pkglist = hy_query_run (query);
+  for (guint i = 0; i < pkglist->len; i++)
+    {
+      auto pkg = static_cast<DnfPackage *> (g_ptr_array_index (pkglist, i));
+      os_add_package_info_to_builder (pkg, builder, id);
+    }
+}
+
+static void
+search_packages_by_filter (HyQuery query, GVariantBuilder *builder, const gchar *const *names,
+                           int *keynames, const gchar *id)
+{
+  hy_query_clear (query);
+  for (guint j = 0; keynames[j] != -1; j++)
+    {
+      for (guint i = 0; names[i] != NULL; i++)
+        {
+          if (!hy_is_glob_pattern (names[i]))
+            {
+              hy_query_filter (query, keynames[j], HY_EQ | HY_ICASE, names[i]);
+            }
+          else
+            {
+              hy_query_filter (query, keynames[j], HY_GLOB | HY_ICASE, names[i]);
+            }
+        }
+    }
+  query_results_to_builder (query, builder, id);
+
+  hy_query_clear (query);
+  for (guint j = 0; keynames[j] != -1; j++)
+    {
+      for (guint i = 0; names[i] != NULL; i++)
+        {
+          if (!hy_is_glob_pattern (names[i]))
+            {
+              hy_query_filter (query, keynames[j], HY_SUBSTR | HY_ICASE, names[i]);
+            }
+          else
+            {
+              hy_query_filter (query, keynames[j], HY_GLOB | HY_ICASE, names[i]);
+            }
+        }
+    }
+  query_results_to_builder (query, builder, id);
+}
+
+static gboolean
+os_handle_search (RPMOSTreeOS *interface, GDBusMethodInvocation *invocation,
+                  const gchar *const *names)
+{
+  GError *local_error = NULL;
+  g_autoptr (GCancellable) cancellable = NULL;
+  g_autoptr (DnfContext) dnfctx = NULL;
+
+  sd_journal_print (LOG_INFO, "Handling Search for caller %s",
+                    g_dbus_method_invocation_get_sender (invocation));
+
+  dnfctx = os_create_dnf_context_simple (interface, TRUE, cancellable, &local_error);
+  if (dnfctx == NULL)
+    return os_throw_dbus_invocation_error (invocation, &local_error);
+
+  hy_autoquery HyQuery query = hy_query_create (dnf_context_get_sack (dnfctx));
+
+  GVariantBuilder builder;
+  g_variant_builder_init (&builder, (const GVariantType *)"aa{sv}");
+
+  int keynames_a[3] = { HY_PKG_NAME, HY_PKG_SUMMARY, -1 };
+  search_packages_by_filter (query, &builder, names, keynames_a, "Name & Summary");
+
+  int keynames_b[2] = { HY_PKG_NAME, -1 };
+  search_packages_by_filter (query, &builder, names, keynames_b, "Name");
+
+  int keynames_c[2] = { HY_PKG_SUMMARY, -1 };
+  search_packages_by_filter (query, &builder, names, keynames_c, "Summary");
+
+  GVariant *pkgs_result = g_variant_builder_end (&builder);
+  g_dbus_method_invocation_return_value (invocation, g_variant_new ("(@aa{sv})", pkgs_result));
```

Comment:
Also one important thing here...there's a limit on maximum message size (from [the DBus specification](https://dbus.freedesktop.org/doc/dbus-specification.html)):

> The maximum length of a message, including header, header alignment padding, and body is 2 to the 27th power or 134217728 (128 MiB). Implementations must not send or accept messages exceeding this size. 

Have you looked at how large the message can get with a search like "e" that would match almost every package?

We may want to cap the results at 50 or something for each section?

Reply:
Searching "e" results in an error, likely for the reason you have stated. I've decided to limit the search results to 50 in the API while also moving the duplicate removal from the client end to the API since it makes more sense for it to be there.